test/multidc/inter_dc_zmq_SUITE.erl

```diff
@@ -0,0 +1,453 @@
+        {ok, SubSocket} = chumak:socket(sub),
+        chumak:subscribe(SubSocket, <<>>),
+        {ok, _SocketPid} = chumak:connect(SubSocket, tcp, "localhost", 15551),
+        {ok, MM1} = chumak:recv(SubSocket),
+        ct:log("Received test message ~p", [MM1]),
+
+
+        {ok, SubSocket2} = chumak:socket(sub),
+        ct:log("Socket 1: ~p   Socket 2: ~p",[SubSocket, SubSocket2]),
+
+        ct:log("Subscribing to ~p",[P1Bin]),
+        chumak:subscribe(SubSocket2, P1Bin),
+
+        ct:log("Subscribing to ~p",[P3Bin]),
+        chumak:subscribe(SubSocket2, P3Bin),
+
+        {ok, _} = chumak:connect(SubSocket2, tcp, "localhost", 15551),
+        timer:sleep(50),
+        Self ! more,
+
+        {ok, _MMM1} = chumak:recv(SubSocket2),
+        {ok, _MMM2} = chumak:recv(SubSocket2),
+
+        %% subscriber_timeout!
+%%        {ok, _} = chumak:recv(SubSocket2),
+%%        logger:warning("2"),
+
+        ok = inter_dc_utils:close_socket(SubSocket),
+        Self ! finish
+          end),
+
+    timer:sleep(50),
+
+    % internal API, broadcast allows only for transactions
+    ok = gen_server:call(Pid, {publish, M0}),
+
+    receive more -> ok after 100 -> throw(subscriber_timeout) end,
+
+    ct:log("Publishing ~p",[M1]),
+    ok = gen_server:call(Pid, {publish, M1}),
+    ct:log("Publishing ~p",[M2]),
+    ok = gen_server:call(Pid, {publish, M2}),
+    ct:log("Publishing ~p",[M3]),
+    ok = gen_server:call(Pid, {publish, M3}),
+    ct:log("Publishing ~p",[M4]),
+    ok = gen_server:call(Pid, {publish, M4}),
+
+    receive finish -> ok after 200 -> throw(subscriber_timeout) end,
+    gen_server:stop(Pid),
+    ok.
+
+router_multiple_req(_Config) ->
+    % start local router
+    _RouterPid = spawn(fun() ->
+        {ok, RouterSocket} = chumak:socket(router),
+        {ok, _BindPid} = chumak:bind(RouterSocket, tcp, "0.0.0.0", 15556),
+
+        CaseCheck = fun() ->
+            {ok, [Id, <<>>, <<"ping">>]} = chumak:recv_multipart(RouterSocket),
+            chumak:send_multipart(RouterSocket, [Id, <<>>, <<"pong">>])
+                    end,
+
+        CaseCheck(),
+        CaseCheck(),
+        inter_dc_utils:close_socket(RouterSocket)
+        end),
+
+    timer:sleep(50),
+
+    WorkerLoop = fun(Socket, _Id, _Parent) ->
+        spawn_link(fun() ->
+            chumak:send(Socket, <<"ping">>),
+            {ok, Message} = chumak:recv(Socket),
+            ct:log("~p Message received from router async ~p", [_Id, Message])
+                   end
+        )
+                 end,
+
+    StartWorker = fun () ->
+        Parent = self(),
+        spawn_link(
+            fun() ->
+                Id = atom_to_list(node())++pid_to_list(self()),
+                {ok, Socket} = chumak:socket(req, Id),
+                {ok, _PeerPid} = chumak:connect(Socket, tcp, "localhost", 15556),
+                WorkerLoop(Socket, Id, Parent)
+            end
+        ) end,
+
+
+    StartWorker(),
+    StartWorker(),
+
+    timer:sleep(50),
```

Comment:
Is this deterministic and stable? Wouldn't it be better to wait for a specific reply?

Reply:
It's a bit hard to get rid of all sleeps with sockets, but I tried to change most of the sleeps to message passing.